### ISSUE TITLE
XM Track/trk Calculation

### ISSUE THREAD

**Issue by NoSuck** [libxmp/libxmp#638]
This is more of a question than an issue, but does anyone know why XM track counts are calculated [like this](https://github.com/libxmp/libxmp/blob/master/src/loaders/xm_load.c?plain=1#L855)...

```c
mod->trk = mod->chn * mod->pat + 1;
```

... instead of like this:

```c
mod->trk = mod->chn * (mod->pat + 1);
```

Most other formats adhere to `trk == pat * chn` (with `pat + 1` above being a precalculation).  However, XM does not for some reason.  It's not a bug *per se*, but it starts to matter when you manipulate track data.  (As it turns out, `trk == pat * chn` is a very convenient assumption to be able to make.)

I hope everyone's doing well.  Any insight would be appreciated.

**Comment by AliceLR** [comment]
The extra track is used to support blank patterns—see `load_patterns` line 340. When an order with a number >= the pattern count is encountered, this blank pattern is used. Only one track needs to be allocated for it.

**Comment by NoSuck** [comment]
Thank you.

Does XM's blank pattern have to be different from IT's `+++` (marker) or `---` (end) patterns?  IT does not seem to have a dangling `trk` for those.

**Comment by AliceLR** [comment]
No, they're completely different—missing XM patterns are treated like a complete blank pattern, more like STM than IT.

(Exception: old Modplug XMs can contain IT markers.)

**Comment by NoSuck** [comment]
I see.  Thank you again for the insight, @AliceLR.  I do not track in XM and was unaware of its blank-pattern feature.

Can anyone provide documentation or an example?  The closest I found was [MilkyTracker's `XM.TXT` successor mentioning “empty” patterns](https://github.com/milkytracker/MilkyTracker/blob/master/resources/reference/xm-form.txt?plain#L151,L152).  (Likewise for the original `XM.TXT` from [ft208](https://files.scene.org/get/resources/music/trackers/ft208.zip).)

**Comment by AliceLR** [comment]
Belated (sorry), but I don't know of any *official* documentation that brings this up. This is a behavioral quirk I discovered working on MikMod several years ago: https://github.com/sezero/mikmod/issues/8

(Since libxmp can detect MPT 1.16 XMs now, maybe the stupid marker hack is worth looking at again?)
